File: Massecuite.py
# Massecuite class for vacuum pan calculations.
# BPR regression sourced from Birkett fig 12.14 in his Nicholl's Class Notes (°F). Valid range: ml_purity 30–100.
# Below ml_purity 60 is extrapolated — use with caution.

import functools
import numpy as np
from SteamStream import EvaporatorSteam
from sugar_stream_properties import specific_gravity, get_cp

# --- BPR regression built once at import ---
_PURITIES = [100,  90,   80,   70,   60  ]
_Y1       = [9.9,  11,   12.5, 14.1, 16.3]   # BPR (°F) at T = 130°F
_Y2       = [14.6, 16,   18,   20.3, 22.8]   # BPR (°F) at T = 185°F
_T1, _T2  = 130, 185

def _build_bpr_regression():
    slopes, bs = [], []
    for y1, y2 in zip(_Y1, _Y2):
        m = (y2 - y1) / (_T2 - _T1)
        slopes.append(m)
        bs.append(y2 - m * _T2)
    return (np.poly1d(np.polyfit(_PURITIES, slopes, 1)),
            np.poly1d(np.polyfit(_PURITIES, bs,     1)))

_slope_poly, _b_poly = _build_bpr_regression()

# --- Predicted mother liquor purity from massecuite purity (Birkett correlation) ---
# Birkett found crystal yield (% on brix) regresses linearly against massecuite purity;
# mother liquor purity is then back-calculated from that crystal yield via the standard
# identity CY = (Pm - Pml) / (1 - Pml/100), solved for Pml.
# DISPLAY AID ONLY: an approximate check against the ml_purity actually supplied to
# Pan/Massecuite (lab analysis or process knowledge) -- not a substitute for it.

def predict_mother_liquor_purity(masse_purity, use_birkett=True, slope=None, intercept=None):
    """Predict mother liquor purity (%) from massecuite purity (%).

    use_birkett=True (default) uses Birkett's fitted crystal-yield-vs-massecuite-purity
    slope/intercept (0.7609, -11.693). Pass use_birkett=False with your own slope/intercept
    to use a different crystal-yield correlation instead.
    """
    if use_birkett:
        m, b = 0.7609, -11.693
    else:
        m, b = float(slope), float(intercept)
    crystal_yield = m * masse_purity + b
    return (masse_purity - crystal_yield) / (1 - crystal_yield / 100)


class Massecuite:
    """
    Represents a massecuite in a vacuum pan.

    ml_purity       : mother liquor apparent purity (10–100)
                      Note: BPR data exists for 60–100; below 60 is extrapolated.
    masse_purity    : overall massecuite apparent purity
    masse_brix      : overall massecuite Brix
    inches_vacuum   : vapor space vacuum in inches Hg
    supersaturation : target supersaturation coefficient (e.g. 1.05) — boiling mode
    head_ft         : massecuite depth in the pan (ft) — required in boiling mode
    flow_lb_hr      : massecuite flow rate (lb/hr) — optional; unlocks flow-based properties
    temp_F          : explicit massecuite temperature (°F) — set-temperature mode

    Two modes — give exactly one of supersaturation / temp_F:

    Boiling mode (supersaturation given): the massecuite is boiling in a pan,
    and its temperature is solved from vacuum, static head, and the BPR
    regression. All boiling properties (water bp, BPR at surface/head) apply.

    Set-temperature mode (temp_F given): the massecuite is off-boiling
    (crystallizer, reheater, transport) and its temperature is imposed by
    heat exchange. massecuite_temp returns temp_F directly; supersaturation
    is None and the boiling-solve properties raise, since BPR-based
    supersaturation is only meaningful at boiling equilibrium.

    Solve results are cached on first access — iterations run once per instance.
    """

    PURITY_MIN = 20
    PURITY_MAX = 100

    def __init__(self, ml_purity, masse_purity, masse_brix,
                 inches_vacuum, supersaturation=None, head_ft=None,
                 flow_lb_hr=None, temp_F=None):
        if not (self.PURITY_MIN <= ml_purity <= self.PURITY_MAX):
            raise ValueError(
                f"Mother liquor purity {ml_purity} is outside the supported range "
                f"({self.PURITY_MIN}–{self.PURITY_MAX})."
            )
        if masse_purity < ml_purity:
            raise ValueError(
                f"Massecuite purity ({masse_purity}) cannot be less than "
                f"mother liquor purity ({ml_purity})."
            )
        if (supersaturation is None) == (temp_F is None):
            raise ValueError(
                "Give exactly one of supersaturation (boiling mode) "
                "or temp_F (set-temperature mode)."
            )
        if supersaturation is not None and head_ft is None:
            raise ValueError("head_ft is required in boiling (supersaturation) mode.")
        self.ml_purity       = ml_purity
        self.masse_purity    = masse_purity
        self.masse_brix      = masse_brix
        self.inches_vacuum   = inches_vacuum
        self.supersaturation = supersaturation
        self.head_ft         = head_ft
        self.flow_lb_hr      = flow_lb_hr
        self.temp_F          = temp_F

    def copy(self, **changes):
        """
        Return a new Massecuite with the same inputs, overriding any given
        as keyword arguments:

            b_masse = a_masse.copy(masse_brix=93)

        Overriding temp_F switches the copy to set-temperature mode (the old
        supersaturation is dropped), and overriding supersaturation switches
        back to boiling mode — so a pan massecuite can flow into a
        crystallizer with:

            out = pan_masse.copy(temp_F=120, ml_purity=64)

        Goes back through __init__, so validation re-runs and the new
        instance solves fresh (no stale cached results carry over).
        """
        params = {
            'ml_purity':       self.ml_purity,
            'masse_purity':    self.masse_purity,
            'masse_brix':      self.masse_brix,
            'inches_vacuum':   self.inches_vacuum,
            'supersaturation': self.supersaturation,
            'head_ft':         self.head_ft,
            'flow_lb_hr':      self.flow_lb_hr,
            'temp_F':          self.temp_F,
        }
        if changes.get('temp_F') is not None:
            params['supersaturation'] = None
        if changes.get('supersaturation') is not None:
            params['temp_F'] = None
        params.update(changes)
        return type(self)(**params)

    # ------------------------------------------------------------------
    # Private helpers (temperature-parameterized, used during iteration)
    # ------------------------------------------------------------------

    def _density_at(self):
        return specific_gravity(self.masse_brix) * 62.4

    def _require_boiling(self):
        if self.temp_F is not None:
            raise AttributeError(
                "This property needs the boiling solve, but this massecuite was "
                "built with an explicit temp_F (set-temperature mode) — BPR-based "
                "supersaturation is only meaningful at boiling equilibrium. "
                "Use copy(supersaturation=...) if it is back in a pan."
            )

    def _sat_bpr_at(self, temp_F):
        """Saturation BPR (°F) at a given temperature — used inside solve loops."""
        return float(_slope_poly(self.ml_purity)) * temp_F + float(_b_poly(self.ml_purity))

    # ------------------------------------------------------------------
    # Vapor space pressure
    # ------------------------------------------------------------------

    @property
    def vapor_pressure_psia(self):
        """Vapor space absolute pressure (psia)."""
        return 14.696 - self.inches_vacuum * 0.491154

    # ------------------------------------------------------------------
    # Cached solves (run once, stored in instance __dict__ by cached_property)
    # ------------------------------------------------------------------

    @functools.cached_property
    def _surface_solve(self):
        """Solve for massecuite temp at surface (head = 0). Returns (T, water_bp)."""
        self._require_boiling()
        water_bp = EvaporatorSteam(self.vapor_pressure_psia).sat_temp_deg_F
        T = water_bp + 20
        for _ in range(100):
            T_new = water_bp + self.supersaturation * self._sat_bpr_at(T)
            if abs(T_new - T) < 1e-6:
                return T_new, water_bp
            T = T_new
        raise RuntimeError("Massecuite surface solve did not converge")

    @functools.cached_property
    def _head_solve(self):
        """Solve for massecuite temp at specified head depth. Returns (T, water_bp_at_head)."""
        self._require_boiling()
        P_vapor = self.vapor_pressure_psia
        T = EvaporatorSteam(P_vapor).sat_temp_deg_F + 20
        for _ in range(100):
            delta_P       = self._density_at() * self.head_ft / 144
            water_bp_head = EvaporatorSteam(P_vapor + delta_P).sat_temp_deg_F
            T_new         = water_bp_head + self.supersaturation * self._sat_bpr_at(T)
            if abs(T_new - T) < 1e-6:
                return T_new, water_bp_head
            T = T_new
        raise RuntimeError("Massecuite head solve did not converge")

    # ------------------------------------------------------------------
    # Surface condition (head = 0)
    # ------------------------------------------------------------------

    @property
    def water_bp_surface(self):
        """Water boiling point at the vapor space — no static head (°F)."""
        return self._surface_solve[1]

    @property
    def massecuite_temp_surface(self):
        """Massecuite temperature to reach target SS at the surface (°F)."""
        return self._surface_solve[0]

    @property
    def bpr_at_surface(self):
        """BPR at the surface condition (°F)."""
        return self.massecuite_temp_surface - self.water_bp_surface

    # ------------------------------------------------------------------
    # Head condition (at specified depth)
    # ------------------------------------------------------------------

    @property
    def water_bp_at_head(self):
        """Water boiling point at the specified pan depth, including static head (°F)."""
        return self._head_solve[1]

    @property
    def massecuite_temp(self):
        """Massecuite temperature (°F) — the given temp_F in set-temperature mode,
        otherwise solved to reach target SS at the specified head depth."""
        if self.temp_F is not None:
            return self.temp_F
        return self._head_solve[0]

    @property
    def bpr_at_head(self):
        """BPR at the specified head depth (°F)."""
        return self.massecuite_temp - self.water_bp_at_head

    # ------------------------------------------------------------------
    # Physical properties at the converged head temperature
    # ------------------------------------------------------------------

    @property
    def density(self):
        """Massecuite density."""
        return self._density_at()

    @property
    def saturation_bpr(self):
        """Saturation BPR at the converged head temperature — SS = 1.0 reference (°F)."""
        return self._sat_bpr_at(self.massecuite_temp)
    
    @property
    def specific_heat(self):
        return get_cp(self.masse_brix)

    # ------------------------------------------------------------------
    # Composition
    # ------------------------------------------------------------------

    @property
    def crystal_content(self):
        """
        Crystal content (% by mass of massecuite).
            C = masse_brix × (masse_purity - ml_purity) / (100 - ml_purity)
        """
        return self.masse_brix * (self.masse_purity - self.ml_purity) / (100 - self.ml_purity)

    @property
    def mother_liquor_brix(self):
        """
        Mother liquor Brix — derived from overall Brix and crystal content via water balance:
            B_ml = 100 - 100 × (100 - masse_brix) / (100 - crystal_content)
        """
        return 100 - 100 * (100 - self.masse_brix) / (100 - self.crystal_content)

    @property
    def crystal_yield_pct_brix(self):
        """
        Crystal yield (%) — fraction of sucrose crystallized relative to what is
        theoretically achievable at this mother liquor purity:
            yield = (masse_purity - ml_purity) / (100 - ml_purity) × 100
        """
        return (self.masse_purity - self.ml_purity) / (100 - self.ml_purity) * 100

    # ------------------------------------------------------------------
    # Flow-based properties (require flow_lb_hr to be set)
    # ------------------------------------------------------------------

    def _check_flow(self):
        if self.flow_lb_hr is None:
            raise AttributeError(
                "flow_lb_hr is not set — pass it to the constructor or assign it directly."
            )

    @property
    def solids_flow(self):
        """Total dissolved + crystal solids flow (lb/hr).  = masse_brix/100 × flow"""
        self._check_flow()
        return self.masse_brix / 100 * self.flow_lb_hr

    @property
    def pol_flow(self):
        """Sucrose (pol) flow (lb/hr).  = masse_purity × masse_brix / 10000 × flow"""
        self._check_flow()
        return self.masse_purity * self.masse_brix / 10000 * self.flow_lb_hr
    
    @property
    def cu_ft_hr(self):
        return self.flow_lb_hr / self.density

    # ------------------------------------------------------------------
    # Dunder / display
    # ------------------------------------------------------------------

    def __repr__(self):
        flow_str = f", flow_lb_hr={self.flow_lb_hr:,.0f}" if self.flow_lb_hr is not None else ""
        mode_str = (f"temp_F={self.temp_F}" if self.temp_F is not None else
                    f"inches_vacuum={self.inches_vacuum}, "
                    f"supersaturation={self.supersaturation}, head_ft={self.head_ft}")
        return (
            f"Massecuite(ml_purity={self.ml_purity}, masse_purity={self.masse_purity}, "
            f"masse_brix={self.masse_brix}, {mode_str}{flow_str})"
        )

    def properties(self):
        d = {
            'ml_purity':                     self.ml_purity,
            'masse_purity':                  self.masse_purity,
            'masse_brix':                    self.masse_brix,
            'crystal_content_pct':           self.crystal_content,
            'mother_liquor_brix':            self.mother_liquor_brix,
            'crystal_yield_pct_brix_pct':    self.crystal_yield_pct_brix,
            'density_lb_ft3':                self.density,
            'massecuite_temp':               self.massecuite_temp,
            'sat_bpr':                       self.saturation_bpr,
        }
        if self.temp_F is None:   # boiling mode — full solve available
            d.update({
                'inches_vacuum':             self.inches_vacuum,
                'vapor_pressure_psia':       self.vapor_pressure_psia,
                'supersaturation':           self.supersaturation,
                'head_ft':                   self.head_ft,
                'water_bp_surface':          self.water_bp_surface,
                'massecuite_temp_surf':      self.massecuite_temp_surface,
                'bpr_at_surface':            self.bpr_at_surface,
                'water_bp_at_head':          self.water_bp_at_head,
                'bpr_at_head':               self.bpr_at_head,
            })
        if self.flow_lb_hr is not None:
            d['flow_lb_hr']    = self.flow_lb_hr
            d['solids_flow']   = self.solids_flow
            d['pol_flow']      = self.pol_flow
        return d

    def display_properties(self):
        props = self.properties()
        print("Units: T(°F), BPR(°F), density(lb/ft³), flow(lb/hr), purity/brix/yield(%)")
        for k, v in props.items():
            print(f"  {k:<29}: {v:.3f}")


if __name__ == "__main__":
    import time

    masse = Massecuite(ml_purity=70, masse_purity=90, masse_brix=92,
                       inches_vacuum=23.5, supersaturation=1.2, head_ft=2,
                       flow_lb_hr=500_000)
    print(masse)
    print()
    t0 = time.perf_counter()
    masse.display_properties()
    print(masse)
    print(f"\nchanging temp test\n")
    masse.temp_F = 160
    masse.display_properties()
    elapsed_ms = (time.perf_counter() - t0) * 1000
    print(f"\n  solved in {elapsed_ms:.3f} ms")

File: web/src/massecuite.js
// Port of Massecuite.py — massecuite state in a vacuum pan (or off-boiling in
// a crystallizer/reheater/transport). BPR regression sourced from Birkett fig
// 12.14 in his Nicholl's Class Notes (deg F). Valid range: ml_purity 30-100,
// below 60 is extrapolated.

(function (root) {
  'use strict';

  const SugarStream = (typeof module !== 'undefined' && module.exports) ? require('./sugar_stream.js') : root.SugarStream;
  const { satSteamTemp, getCp, specificGravity } = SugarStream._properties;

  // --- BPR regression, built once at module load (mirrors the Python
  // module-level _build_bpr_regression() call) ---
  const _PURITIES = [100, 90, 80, 70, 60];
  const _Y1 = [9.9, 11, 12.5, 14.1, 16.3];   // BPR (F) at T = 130F
  const _Y2 = [14.6, 16, 18, 20.3, 22.8];    // BPR (F) at T = 185F
  const _T1 = 130, _T2 = 185;

  // np.polyfit(x, y, 1) — least-squares line, returned as [slope, intercept].
  function linearFit(xs, ys) {
    const n = xs.length;
    let sx = 0, sy = 0, sxx = 0, sxy = 0;
    for (let i = 0; i < n; i++) {
      sx += xs[i]; sy += ys[i]; sxx += xs[i] * xs[i]; sxy += xs[i] * ys[i];
    }
    const slope = (n * sxy - sx * sy) / (n * sxx - sx * sx);
    const intercept = (sy - slope * sx) / n;
    return [slope, intercept];
  }

  function _buildBprRegression() {
    const slopes = [], bs = [];
    for (let i = 0; i < _PURITIES.length; i++) {
      const m = (_Y2[i] - _Y1[i]) / (_T2 - _T1);
      slopes.push(m);
      bs.push(_Y2[i] - m * _T2);
    }
    return [linearFit(_PURITIES, slopes), linearFit(_PURITIES, bs)];
  }

  const [_slopeCoef, _bCoef] = _buildBprRegression();
  const _slopePoly = (purity) => _slopeCoef[0] * purity + _slopeCoef[1];
  const _bPoly = (purity) => _bCoef[0] * purity + _bCoef[1];

  const PURITY_MIN = 20;
  const PURITY_MAX = 100;

  class Massecuite {
    constructor({ ml_purity, masse_purity, masse_brix, inches_vacuum,
                  supersaturation = null, head_ft = null,
                  flow_lb_hr = null, temp_F = null } = {}) {
      if (!(ml_purity >= PURITY_MIN && ml_purity <= PURITY_MAX)) {
        throw new Error(`Mother liquor purity ${ml_purity} is outside the supported range (${PURITY_MIN}-${PURITY_MAX}).`);
      }
      if (masse_purity < ml_purity) {
        throw new Error(`Massecuite purity (${masse_purity}) cannot be less than mother liquor purity (${ml_purity}).`);
      }
      const ssGiven = supersaturation !== null && supersaturation !== undefined;
      const tempGiven = temp_F !== null && temp_F !== undefined;
      if (ssGiven === tempGiven) {
        throw new Error('Give exactly one of supersaturation (boiling mode) or temp_F (set-temperature mode).');
      }
      if (ssGiven && (head_ft === null || head_ft === undefined)) {
        throw new Error('head_ft is required in boiling (supersaturation) mode.');
      }
      this.ml_purity = ml_purity;
      this.masse_purity = masse_purity;
      this.masse_brix = masse_brix;
      this.inches_vacuum = inches_vacuum;
      this.supersaturation = ssGiven ? supersaturation : null;
      this.head_ft = head_ft;
      this.flow_lb_hr = flow_lb_hr;
      this.temp_F = tempGiven ? temp_F : null;

      this._surfaceSolveCache = null;
      this._headSolveCache = null;
    }

    copy(changes = {}) {
      const params = {
        ml_purity: this.ml_purity,
        masse_purity: this.masse_purity,
        masse_brix: this.masse_brix,
        inches_vacuum: this.inches_vacuum,
        supersaturation: this.supersaturation,
        head_ft: this.head_ft,
        flow_lb_hr: this.flow_lb_hr,
        temp_F: this.temp_F,
      };
      if (changes.temp_F !== undefined && changes.temp_F !== null) {
        params.supersaturation = null;
      }
      if (changes.supersaturation !== undefined && changes.supersaturation !== null) {
        params.temp_F = null;
      }
      Object.assign(params, changes);
      return new Massecuite(params);
    }

    // ------------------------------------------------------------------
    // Private helpers (temperature-parameterized, used during iteration)
    // ------------------------------------------------------------------

    _densityAt() {
      return specificGravity(this.masse_brix) * 62.4;
    }

    _requireBoiling() {
      if (this.temp_F !== null) {
        throw new Error(
          'This property needs the boiling solve, but this massecuite was built with an ' +
          'explicit temp_F (set-temperature mode) -- BPR-based supersaturation is only ' +
          'meaningful at boiling equilibrium. Use copy({supersaturation: ...}) if it is back in a pan.'
        );
      }
    }

    _satBprAt(temp_F) {
      return _slopePoly(this.ml_purity) * temp_F + _bPoly(this.ml_purity);
    }

    // ------------------------------------------------------------------
    // Vapor space pressure
    // ------------------------------------------------------------------

    get vapor_pressure_psia() {
      return 14.696 - this.inches_vacuum * 0.491154;
    }

    // ------------------------------------------------------------------
    // Cached solves (run once, memoized like Python's functools.cached_property)
    // ------------------------------------------------------------------

    _surfaceSolve() {
      if (this._surfaceSolveCache) return this._surfaceSolveCache;
      this._requireBoiling();
      const water_bp = satSteamTemp(this.vapor_pressure_psia);
      let T = water_bp + 20;
      for (let i = 0; i < 100; i++) {
        const T_new = water_bp + this.supersaturation * this._satBprAt(T);
        if (Math.abs(T_new - T) < 1e-6) {
          this._surfaceSolveCache = [T_new, water_bp];
          return this._surfaceSolveCache;
        }
        T = T_new;
      }
      throw new Error('Massecuite surface solve did not converge');
    }

    _headSolve() {
      if (this._headSolveCache) return this._headSolveCache;
      this._requireBoiling();
      const P_vapor = this.vapor_pressure_psia;
      let T = satSteamTemp(P_vapor) + 20;
      for (let i = 0; i < 100; i++) {
        const delta_P = this._densityAt() * this.head_ft / 144;
        const water_bp_head = satSteamTemp(P_vapor + delta_P);
        const T_new = water_bp_head + this.supersaturation * this._satBprAt(T);
        if (Math.abs(T_new - T) < 1e-6) {
          this._headSolveCache = [T_new, water_bp_head];
          return this._headSolveCache;
        }
        T = T_new;
      }
      throw new Error('Massecuite head solve did not converge');
    }

    // ------------------------------------------------------------------
    // Surface condition (head = 0)
    // ------------------------------------------------------------------

    get water_bp_surface() { return this._surfaceSolve()[1]; }
    get massecuite_temp_surface() { return this._surfaceSolve()[0]; }
    get bpr_at_surface() { return this.massecuite_temp_surface - this.water_bp_surface; }

    // ------------------------------------------------------------------
    // Head condition (at specified depth)
    // ------------------------------------------------------------------

    get water_bp_at_head() { return this._headSolve()[1]; }

    get massecuite_temp() {
      if (this.temp_F !== null) return this.temp_F;
      return this._headSolve()[0];
    }

    get bpr_at_head() { return this.massecuite_temp - this.water_bp_at_head; }

    // ------------------------------------------------------------------
    // Physical properties at the converged head temperature
    // ------------------------------------------------------------------

    get density() { return this._densityAt(); }

    get saturation_bpr() { return this._satBprAt(this.massecuite_temp); }

    get specific_heat() { return getCp(this.masse_brix); }

    // ------------------------------------------------------------------
    // Composition
    // ------------------------------------------------------------------

    get crystal_content() {
      return this.masse_brix * (this.masse_purity - this.ml_purity) / (100 - this.ml_purity);
    }

    get mother_liquor_brix() {
      return 100 - 100 * (100 - this.masse_brix) / (100 - this.crystal_content);
    }

    get crystal_yield_pct_brix() {
      return (this.masse_purity - this.ml_purity) / (100 - this.ml_purity) * 100;
    }

    // ------------------------------------------------------------------
    // Flow-based properties (require flow_lb_hr to be set)
    // ------------------------------------------------------------------

    _checkFlow() {
      if (this.flow_lb_hr === null || this.flow_lb_hr === undefined) {
        throw new Error('flow_lb_hr is not set -- pass it to the constructor or assign it directly.');
      }
    }

    get solids_flow() {
      this._checkFlow();
      return this.masse_brix / 100 * this.flow_lb_hr;
    }

    get pol_flow() {
      this._checkFlow();
      return this.masse_purity * this.masse_brix / 10000 * this.flow_lb_hr;
    }

    get cu_ft_hr() {
      return this.flow_lb_hr / this.density;
    }

    // ------------------------------------------------------------------
    // Display
    // ------------------------------------------------------------------

    properties() {
      const d = {
        ml_purity: this.ml_purity,
        masse_purity: this.masse_purity,
        masse_brix: this.masse_brix,
        crystal_content_pct: this.crystal_content,
        mother_liquor_brix: this.mother_liquor_brix,
        crystal_yield_pct_brix_pct: this.crystal_yield_pct_brix,
        density_lb_ft3: this.density,
        massecuite_temp: this.massecuite_temp,
        sat_bpr: this.saturation_bpr,
      };
      if (this.temp_F === null) {
        Object.assign(d, {
          inches_vacuum: this.inches_vacuum,
          vapor_pressure_psia: this.vapor_pressure_psia,
          supersaturation: this.supersaturation,
          head_ft: this.head_ft,
          water_bp_surface: this.water_bp_surface,
          massecuite_temp_surf: this.massecuite_temp_surface,
          bpr_at_surface: this.bpr_at_surface,
          water_bp_at_head: this.water_bp_at_head,
          bpr_at_head: this.bpr_at_head,
        });
      }
      if (this.flow_lb_hr !== null && this.flow_lb_hr !== undefined) {
        d.flow_lb_hr = this.flow_lb_hr;
        d.solids_flow = this.solids_flow;
        d.pol_flow = this.pol_flow;
      }
      return d;
    }
  }

  Massecuite._properties = { linearFit, _slopePoly, _bPoly };

  if (typeof module !== 'undefined' && module.exports) {
    module.exports = Massecuite;
  } else {
    root.Massecuite = Massecuite;
  }
})(typeof window !== 'undefined' ? window : globalThis);
